### PR TITLE
created an orbital solver as an improved replacement to body lock solver

### PR DESCRIPTION
**Pull request by Ecnassianer**
This pull request creates an orbital solver to replace the body locked solver. The Orbital Solver fixes numerous issues with using the Body Locked solver, rotation issues, bad messaging, non-working features, and more.

Details are available at this proposal:
https://github.com/Microsoft/MixedRealityToolkit-Unity/issues/2250

### CODE REVIEW

**Comment by SimonDarksideJ** [comment]
I agree, at this late stage, this would be better suited to vNext

**Comment by davidkline-ms** [comment]
Consider vNext seconded.

@Ecnassianer, can you retarget this to vNext? Unless @keveleigh feels strongly, I do not see the need to replace the body lock solver for the HoloToolkit.

**Comment by Ecnassianer** [comment]
Sure, I haven't checked into vNext yet. Is there anything I need to do other than point this PR at a new branch?

**Comment by StephenHodgson** [comment]
It's probably not a direct 1:1.

File: Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+        private Quaternion SnapToTetherAngleSteps(Quaternion RotationToSnap)
+        {
+            if (!UseAngleSteppingForWorldOffset)
+            {
+                return RotationToSnap;
+            }
+
+            float stepAngle = 360f / TetherAngleSteps;
+            float overage = solverHandler.TransformTarget.transform.localEulerAngles.y % stepAngle;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs]: not used?

File: Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        public Vector3 LocalOffset
+        {
+            get { return localOffset; }
+            set { localOffset = value; }
+        }
+
+        [Tooltip("XYZ offset for this object in worldspace, best used with the YawOnly orientation. Mixing local and world offsets is not recommended.")]
+        [SerializeField]
+        private Vector3 worldOffset = Vector3.zero;
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs]: Let's add summary tags with the text from the tooltips to match for all properties

File: Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        [Tooltip("TetherAngleSteps is the division of steps this object can tether to. Higher the number, the more snapple steps.")]
+        [Range(2, 24)]
+        [SerializeField]
+        private int tetherAngleSteps = 6;
+
+        public int TetherAngleSteps
+        {
+            get { return tetherAngleSteps;}
+            set { tetherAngleSteps = value; }
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs]: We should be checking the value here to make sure it's within our range above.

```

/// <summary>
/// The division of steps this object can tether to. Higher the number, the more snapple steps.
/// </summary>
public int TetherAngleSteps
{
    get { return tetherAngleSteps; }
    set
    {
        if (value < 2)
        {
            tetherAngleSteps = 2;
        }
        else if (value > 24)
        {
            tetherAngleSteps = 24;
        }
        else
        {
            tetherAngleSteps = value;
        }
    }
}
```

File: Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+namespace HoloToolkit.Unity
+{
+    /// <summary>
+    /// SolverObjectLock provides a solver that follows the TrackedObject/TargetTransform. Adjusting "LerpTime"
+    /// properties changes how quickly the object moves to the TrackedObject/TargetTransform's position.
+    /// </summary>
+    public class SolverOrbital : Solver
+    {
+        #region public enums
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs]: do we really need a region for a single enum?

**Comment by StephenHodgson** [comment]
@Ecnassianer I've already replaced this in vNEXT. You'll see it soon in my next PR

File: Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs [hunk 5]
```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+
+namespace HoloToolkit.Unity
+{
+    /// <summary>
+    /// SolverObjectLock provides a solver that follows the TrackedObject/TargetTransform. Adjusting "LerpTime"
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Scripts/Solvers/SolverOrbital.cs]: This description doesn't really fix anymore.

**Comment by SimonDarksideJ** [comment]
Is this now resolved @Ecnassianer / @StephenHodgson if so can we close this PR as it seems stale

**Comment by keveleigh** [comment]
@SimonDarksideJ Thanks for pinging all the legacy PRs. I've been reviewing these, as I find time, on my end for inclusion in the September release. Feel free to add your own review as well!

This one is not resolved, as the new solver exists in vNext only. This PR is targeting htk_development.

**Comment by SimonDarksideJ** [comment]
In which case this one should probably be closed, as all agreed this should only be in vNext due to its impacts

**Comment by keveleigh** [comment]
After using this during my port, I think it'd be a great improvement for the September release, but I'll close it due to previously expressed opinions.

**Comment by StephenHodgson** [comment]
Idk, I don't see why we can't keep it as long as we mark any breaking changes as obsolete

**Comment by StephenHodgson** [comment]
After going back and reviewing this I would have approved it.

**Comment by Ecnassianer** [comment]
>Idk, I don't see why we can't keep it as long as we mark any breaking changes as obsolete

Better yet, the only change to existing code is a warning message saying to use Orbital Solvers instead. The Body Lock Solvers still work exactly the same as before.

**Pull request merged**